hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteArrayObjectDataInput.java

```diff
@@ -522,10 +577,23 @@ public int readUnsignedShort() throws IOException {
      */
     @Override
     public final String readUTF() throws IOException {
-        if (utfBuffer == null) {
-            utfBuffer = new byte[UTF_BUFFER_SIZE];
+        int charCount = readInt();
+        if (charCount == NULL_ARRAY) {
+            return null;
+        }
+        if (charBuffer == null || charCount > charBuffer.length) {
+            charBuffer = new char[charCount];
+        }
+        byte b;
+        for (int i = 0; i < charCount; i++) {
+            b = readByte();
+            if (b < 0) {
+                charBuffer[i] = Bits.readUtf8Char(this, b);
+            } else {
+                charBuffer[i] = (char) b;
+            }
         }
-        return UTFEncoderDecoder.readUTF(this, utfBuffer);
+        return new String(charBuffer, 0, charCount);
```

Comment:
@jerrinot @serkan-ozal @noctarius can you have a look at this? You know quite a lot about the UTFEncoderDecoder.

What is the reason the UTFEncoderDecoder is dropped?